crates/sui-framework/src/lib.rs

```diff
@@ -146,3 +160,90 @@ pub fn build_move_package(path: &Path, config: BuildConfig) -> SuiResult<Compile
     }*/
     Ok(pkg)
 }
+
+/// Check whether the framework defined by `modules` is compatible with the framework that is
+/// already on-chain (i.e. stored in `object_store`) at `id`.
+///
+/// - Returns `None` if the current package at `id` cannot be loaded, or the compatibility check
+///   fails (This is grounds not to upgrade).
+/// - Panics if the object at `id` can be loaded but is not a package -- this is an invariant
+///   violation.
+/// - Returns the digest of the current framework (and version) if it is equivalent to the new
+///   framework (indicates support for a protocol upgrade without a framework upgrade).
+/// - Returns the digest of the new framework (and version) if it is compatible (indicates
+///   support for a protocol upgrade with a framework upgrade).
+pub async fn compare_system_package<S: ObjectStore>(
```

Comment:
This function is moved from authority.rs to framework.
The only change is the first param becomes ObjectStore instead of self.